subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -99,6 +183,46 @@ lwm2m_engine_set_data_cb_t lwm2m_firmware_get_write_cb(void)
 	return write_cb;
 }
 
+void lwm2m_firmware_set_update_cb(lwm2m_engine_exec_cb_t cb)
+{
+	update_cb = cb;
+}
+
+lwm2m_engine_exec_cb_t lwm2m_firmware_get_update_cb(void)
+{
+	return update_cb;
+}
+
+static int firmware_update_cb(u16_t obj_inst_id)
+{
+	lwm2m_engine_exec_cb_t callback;
+	u8_t state;
+	int ret;
+
+	state = lwm2m_firmware_get_update_state();
+	if (STATE_DOWNLOADED != state) {
+		/* TODO: pass response code to caller, -1 will be shown as
+		 * 4.05 method not allowed in current implmentation */
+		SYS_LOG_ERR("State other than downloaded: %d", state);
+		return -1;
+	}
+
+	lwm2m_firmware_set_update_state(STATE_UPDATING);
+
+	callback = lwm2m_firmware_get_update_cb();
+	if (callback) {
+		ret = callback(obj_inst_id);
+		if (ret < 0) {
+			SYS_LOG_ERR("Failed to update firmware: %d", ret);
+			lwm2m_firmware_set_update_result(
+					RESULT_INTEGRITY_FAILED);
```

Comment:
Good suggestion.
I will map -EINVAL to RESULT_INTEGRITY_FAILED and others to RESULT_UPDATE_FAILED.
From [LwM2M specification](https://goo.gl/5HYazk) figure 29, it seems there are only three acceptable results which are RESULT_SUCCESS, RESULT_INTEGRITY_FAILED and RESULT_UPDATE_FAILED.